pkg/cmd/extension/push/cmd_test.go

```diff
@@ -97,3 +116,42 @@ Pushed %s:latest
 digest: sha256`, localRegistryWasmImageRef))
 	require.Empty(t, stderr, `expected no stderr running [%v]`, c)
 }
+
+// The tests above are unit tests, not end-to-end (e2e) tests. Hence, we use a mock registry instead of a real one.
+func mockRegistryServer() *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		statusCode := 500
+		switch r.Method {
+		case "HEAD":
+			if strings.Index(r.URL.Path, "/v2/getenvoy/sample/blobs") == 0 {
+				statusCode = 404 // pretend it hasn't been uploaded, yet
+			} else if r.URL.Path == "/v2/getenvoy/sample/manifests/latest" {
+				statusCode = 404 // pretend there's no manifest either
+			}
+		case "POST":
+			if r.URL.Path == "/v2/getenvoy/sample/blobs/uploads/" {
+				statusCode = 202 // pretend we processed the data
+				w.Header().Add("Location", "/upload")
+			}
+		case "PUT":
+			if r.URL.Path == "/upload" {
+				statusCode = 200 // Pretend we accepted the blob
+				_ = r.ParseForm()
+				w.Header().Add("Docker-Content-Digest", r.Form.Get("digest"))
+			} else if r.URL.Path == "/v2/getenvoy/sample/manifests/latest" {
+				statusCode = 200 // Pretend we accepted the manifest
+				if raw, err := ioutil.ReadAll(r.Body); err == nil {
+					w.Header().Add("Docker-Content-Digest", "sha256:"+hash(raw))
+				}
```

Comment:
nit looks like you forgot to handle `err != nil` 😄

Reply:
yeah I actually considered it and got lazy. I think I will make error nil a 500. good point!